### PR TITLE
Handle deserialization errors in CLI and JDBC

### PR DESCRIPTION
**Pull request by electrum**

### CODE REVIEW

File: presto-client/src/main/java/com/facebook/presto/client/QueryResults.java [hunk 1]
```diff
@@ -339,4 +342,15 @@ private static Object fixValue(TypeSignature signature, Object value)
...
         }
     }
+
+    private static <T> T safeCast(Class<T> type, Object value)
+    {
+        if (!type.isInstance(value)) {
+            throw new ClientException(format(
+                    "Result data has wrong deserialized type (expected: %s, actual: %s)",
+                    type.getName(),
+                    value.getClass().getName()));
```
Comment (findepi) [inline presto-client/src/main/java/com/facebook/presto/client/QueryResults.java]: Can `value == null` at this point?

File: presto-client/src/main/java/com/facebook/presto/client/StatementClient.java [hunk 2]
```diff
@@ -374,6 +375,13 @@ private RuntimeException requestFailedException(String task, Request request, Js
     {
         gone.set(true);
         if (!response.hasValue()) {
+            if (response.getException() != null) {
+                for (Throwable throwable : getCausalChain(response.getException())) {
```
Comment (findepi) [inline presto-client/src/main/java/com/facebook/presto/client/StatementClient.java]: Why are you discarding exception chain up to (excluding) first `ClientException`? (I'm worried that discarding exceptions will hurt problem diagnosability)

Can it be that `ClientException` is "buried"? 
If so, what about `return new ClientException(...., response.getException())` when causal chain contains a `ClientException`?

Comment (electrum) [reply presto-client/src/main/java/com/facebook/presto/client/StatementClient.java]: This whole fix is really hacky. The `fixValue()` method is called deep inside Jackson serialization, so the `ClientException` will be wrapped several times. We throw `ClientException` when we know that it is the real error that should be presented to the user.

File: presto-client/src/main/java/com/facebook/presto/client/QueryResults.java (hunk 1 above)
Comment (electrum) [reply presto-client/src/main/java/com/facebook/presto/client/QueryResults.java]: We never call this method with null. That's handled at the beginning of `fixValue()`

File: presto-cli/src/main/java/com/facebook/presto/cli/StatusPrinter.java [hunk 3]
```diff
@@ -122,10 +119,13 @@ else if (toUpperCase(key) == 'D') {
                     client.advance();
                 }
                 catch (RuntimeException e) {
-                    log.debug(e, "error printing status");
+                    console.resetScreen();
                     if (debug) {
                         e.printStackTrace(out);
                     }
+                    else {
+                        out.println("ERROR: " + e.getMessage());
```
[the remark this answers is not in the document]
Comment (electrum) [reply presto-cli/src/main/java/com/facebook/presto/cli/StatusPrinter.java]: We weren't printing anything before, so this is strictly better. This error is printed in non-debug mode so we want it to be as friendly as possible. Though I need to rethink if this function should be handling the error at all or if it should just propagate up.

File: presto-client/src/main/java/com/facebook/presto/client/StatementClient.java (hunk 2 above)
Comment (findepi) [reply presto-client/src/main/java/com/facebook/presto/client/StatementClient.java]: Appreciated. Maybe leave this info as a code comment?

File: presto-cli/src/main/java/com/facebook/presto/cli/StatusPrinter.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (findepi) [reply presto-cli/src/main/java/com/facebook/presto/cli/StatusPrinter.java]: I understand. But sole exception message is so often not friendly, especially if we don't know where `e` actually comes from. `e.toString()` is almost always `<class-name>: <message>` and you can manually render `<simple-class-name>: <message>`.

.. or at least `firstNonNull(e.getMessage(), e.toString())`

**Pull request closed without merging**